File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/Lainchan.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.core.site.sites;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.adamantcheese.chan.core.model.orm.Board;
import com.github.adamantcheese.chan.core.site.ChunkDownloaderSiteProperties;
import com.github.adamantcheese.chan.core.site.Site;
import com.github.adamantcheese.chan.core.site.SiteIcon;
import com.github.adamantcheese.chan.core.site.common.CommonSite;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanActions;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanApi;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanCommentParser;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanEndpoints;
import com.github.adamantcheese.chan.core.site.parser.CommentParserType;
import com.github.adamantcheese.model.data.descriptor.ChanDescriptor;

import org.jetbrains.annotations.NotNull;

import okhttp3.HttpUrl;

public class Lainchan extends CommonSite {
    private final ChunkDownloaderSiteProperties chunkDownloaderSiteProperties;
    public static final String SITE_NAME = "Lainchan";

    public static final CommonSiteUrlHandler URL_HANDLER = new CommonSiteUrlHandler() {
        private static final String ROOT = "https://lainchan.org/";

        @Override
        public Class<? extends Site> getSiteClass() {
            return Lainchan.class;
        }

        @Override
        public HttpUrl getUrl() {
            return HttpUrl.parse(ROOT);
        }

        @Override
        public String[] getMediaHosts() {
            return new String[]{ROOT};
        }

        @Override
        public String[] getNames() {
            return new String[]{"lainchan"};
        }

        @Override
        public String desktopUrl(ChanDescriptor chanDescriptor, @Nullable Long postNo) {
            if (chanDescriptor instanceof ChanDescriptor.CatalogDescriptor) {
                return getUrl().newBuilder().addPathSegment(chanDescriptor.boardCode()).toString();
            } else if (chanDescriptor instanceof ChanDescriptor.ThreadDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .addPathSegment("res")
                        .addPathSegment(((ChanDescriptor.ThreadDescriptor) chanDescriptor).getThreadNo() + ".html")
                        .toString();
            } else {
                return getUrl().toString();
            }
        }
    };

    public Lainchan() {
        chunkDownloaderSiteProperties = new ChunkDownloaderSiteProperties(true, true);
    }

    @Override
    public void setup() {
        setEnabled(true);
        setName(SITE_NAME);
        setIcon(SiteIcon.fromFavicon(getImageLoaderV2(), HttpUrl.parse("https://lainchan.org/favicon.ico")));

        setBoards(
                Board.fromSiteNameCode(this, "Programming", "λ"),
                Board.fromSiteNameCode(this, "Do It Yourself", "Δ"),
                Board.fromSiteNameCode(this, "Security", "sec"),
                Board.fromSiteNameCode(this, "Technology", "Ω"),
                Board.fromSiteNameCode(this, "Games and Interactive Media", "inter"),
                Board.fromSiteNameCode(this, "Literature", "lit"),
                Board.fromSiteNameCode(this, "Musical and Audible Media", "music"),
                Board.fromSiteNameCode(this, "Visual Media", "vis"),
                Board.fromSiteNameCode(this, "Humanity", "hum"),
                Board.fromSiteNameCode(this, "Drugs 3.0", "drug"),
                Board.fromSiteNameCode(this, "Consciousness and Dreams", "zzz"),
                Board.fromSiteNameCode(this, "layer", "layer"),
                Board.fromSiteNameCode(this, "Questions and Complaints", "q"),
                Board.fromSiteNameCode(this, "Random", "r"),
                Board.fromSiteNameCode(this, "Lain", "lain"),
                Board.fromSiteNameCode(this, "Culture 15 freshly bumped threads", "culture"),
                Board.fromSiteNameCode(this, "Psychopharmacology 15 freshly bumped threads", "psy"),
                Board.fromSiteNameCode(this, "15 freshly bumped threads", "mega")
        );

        setResolvable(URL_HANDLER);

        setConfig(new CommonConfig() {
            @Override
            public boolean siteFeature(SiteFeature siteFeature) {
                return super.siteFeature(siteFeature) || siteFeature == SiteFeature.POSTING;
            }
        });

        setEndpoints(new VichanEndpoints(this, "https://lainchan.org", "https://lainchan.org"));
        setActions(new VichanActions(this, getOkHttpClient(), getSiteRepository()));
        setApi(new VichanApi(getSiteRepository(), getBoardRepository(), this));
        setParser(new VichanCommentParser(getMockReplyManager()));
    }

    @NotNull
    @Override
    public CommentParserType commentParserType() {
        return CommentParserType.VichanParser;
    }

    @NonNull
    @Override
    public ChunkDownloaderSiteProperties getChunkDownloaderSiteProperties() {
        return chunkDownloaderSiteProperties;
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/lainchan/Lainchan.kt
package com.github.k1rakishou.chan.core.site.sites.lainchan

import com.github.k1rakishou.chan.core.site.SiteActions
import com.github.k1rakishou.chan.core.site.SiteConfiguration
import com.github.k1rakishou.chan.core.site.SiteEndpoints
import com.github.k1rakishou.chan.core.site.SiteUrlHandler
import com.github.k1rakishou.chan.core.site.common.CommonSite
import com.github.k1rakishou.chan.core.site.common.DefaultPostParser
import com.github.k1rakishou.chan.core.site.common.vichan.LainchanCommentParser
import com.github.k1rakishou.chan.core.site.common.vichan.VichanApi
import com.github.k1rakishou.chan.core.site.common.vichan.VichanEndpoints
import com.github.k1rakishou.chan.core.site.limitations.ConstantAttachablesCount
import com.github.k1rakishou.chan.core.site.limitations.ConstantMaxTotalSizeInfo
import com.github.k1rakishou.chan.core.site.limitations.PostingLimitationConfig
import com.github.k1rakishou.chan.core.site.parser.PostParser
import com.github.k1rakishou.chan.core.site.parser.SiteApi
import com.github.k1rakishou.model.data.board.ChanBoard
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.SiteDescriptor
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl

class Lainchan : CommonSite() {
  private val boards by lazy {
    buildList {
      val siteName = descriptor.siteName

      add(ChanBoard.create(BoardDescriptor.create(siteName, "λ"), "Programming"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "Δ"), "Do It Yourself"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "sec"), "Security"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "Ω"), "Technology"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "inter"), "Games and Interactive Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "lit"), "Literature"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "music"), "Musical and Audible Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "vis"), "Visual Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "hum"), "Humanity"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "drug"), "Drugs 3.0"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "zzz"), "Consciousness and Dreams"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "layer"), "layer"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "q"), "Questions and Complaints"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "r"), "Random"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "lain"), "Lain"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "culture"), "Culture 15 freshly bumped threads"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "psy"), "Psychopharmacology 15 freshly bumped threads"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "mega"), "15 freshly bumped threads"))
    }
  }

  override val enabled: Boolean = true
  override val name: String = SITE_NAME
  override val siteIconUrl: HttpUrl = "https://lainchan.org/favicon.ico".toHttpUrl()
  override val commentParserType = SiteConfiguration.CommentParserType.VichanParser
  override val globalSearchType = SiteConfiguration.GlobalSearchType.SearchNotSupported
  override val boardsType: SiteConfiguration.BoardsType = SiteConfiguration.BoardsType.Static
  override val catalogType: SiteConfiguration.CatalogType = SiteConfiguration.CatalogType.Static
  override val chunkedDownloaderConfig by lazy {
    SiteConfiguration.ChunkedDownloaderConfig(
      enabled = true,
      siteSendsCorrectFileSizeInBytes = true
    )
  }
  override val postingLimitationConfig by lazy {
    PostingLimitationConfig(
      postMaxAttachables = ConstantAttachablesCount(3),
      postMaxAttachablesTotalSize = ConstantMaxTotalSizeInfo(75 * (1024 * 1024)) // 75 MB
    )
  }
  override val urlHandler: SiteUrlHandler by lazy { LainchanUrlHandler() }
  override val endpoints: SiteEndpoints by lazy { VichanEndpoints(this, "https://lainchan.org", "https://lainchan.org") }
  override val api: SiteApi by lazy { VichanApi(siteManager, boardManager, this) }
  override val actions: SiteActions by lazy { LainchanActions(this, proxiedOkHttpClient, siteManager, replyManager) }
  override val postParser: PostParser by lazy { DefaultPostParser(LainchanCommentParser(), archivesManager) }
  override val staticBoards: List<ChanBoard> = boards


  override fun hasSiteFeature(siteFeature: SiteConfiguration.SiteFeature): Boolean {
    return super.hasSiteFeature(siteFeature) ||
      siteFeature === SiteConfiguration.SiteFeature.Posting
  }

  class LainchanUrlHandler : CommonSiteUrlHandler() {
    private val ROOT = "https://lainchan.org/"

    override val url: HttpUrl = ROOT.toHttpUrl()
    override val mediaHosts: Array<HttpUrl> = arrayOf(url)

    override fun desktopUrl(chanDescriptor: ChanDescriptor, postNo: Long?, postSubNo: Long?): String? {
      return when (chanDescriptor) {
        is ChanDescriptor.CatalogDescriptor -> {
          url.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .toString()
        }

        is ChanDescriptor.ThreadDescriptor -> {
          url.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .addPathSegment("res")
            .addPathSegment(chanDescriptor.threadNo.toString() + ".html")
            .toString()
        }

        else -> null
      }
    }
  }

  companion object {
    const val SITE_NAME = "Lainchan"
    val SITE_DESCRIPTOR = SiteDescriptor.create(SITE_NAME)
  }
}